File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
* Leetcode Problem - https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
* Author - Shubham Nagaria
* Date - 18th April 2022
Construct Binary Tree from Preorder and Inorder Traversal
**/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* construct(vector<int>& preorder, vector<int>& inorder, int prestart, int preend, int instart, int inend, unordered_map<int,int>&mp){
        if(prestart > preend)
            return NULL;
        TreeNode* node = new TreeNode(preorder[prestart]);
        int indx = mp[preorder[prestart]];
        int x = indx - instart;
        node->left = construct(preorder,inorder, prestart+1, prestart+x,instart,indx-1,mp);
        node->right = construct(preorder,inorder,prestart+x+1,preend, indx+1, inend,mp);
        return node;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int,int>mp;
        int n = inorder.size();
        for(int i = 0; i < n; i++)
            mp[inorder[i]]=i;
        return construct(preorder,inorder,0, n-1, 0,n-1,mp);
        
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def construct(self, ps,pe,ins,ine,preorder,inorder,mp):
        if ps>pe:
            return None
        node = TreeNode(preorder[ps])
        x= mp[preorder[ps]]-ins
        node.left =  self.construct(ps+1,ps+x,ins,ins+x-1,preorder,inorder,mp)
        node.right = self.construct(ps+x+1,pe,ins+x+1,ine,preorder,inorder,mp)
        return node
    
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        mp={}
        for i,j in enumerate(inorder):
            mp[j]=i
        n = len(preorder)-1
        return self.construct(0,n,0,n,preorder,inorder,mp)